### ISSUE TITLE
let_unit_value completely eats variable binding

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#9893]
### Summary

.

### Lint Name

let_unit_value

### Reproducer

I tried this code:

```rust
pub fn main() { let x = (); match x { () => { } } }

```

I saw this happen:
`cargo clippy --fix -- -Aclippy::all -Wclippy::let_unit_value`
```
    Checking clpy v0.1.0 (/tmp/clpy)
warning: failed to automatically apply fixes suggested by rustc to crate `clpy`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0425]: cannot find value `x` in this scope
 --> src/main.rs:1:27
  |
1 | pub fn main() { (); match x { () => { } } }
  |                           ^ not found in this scope

error: aborting due to previous error

For more information about this error, try `rustc --explain E0425`.
Original diagnostics will follow.

warning: this let-binding has unit value
 --> src/main.rs:1:17
  |
1 | pub fn main() { let x = (); match x { () => { } } }
  |                 ^^^^^^^^^^^ help: omit the `let` binding: `();`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#let_unit_value
  = note: requested on the command line with `-W clippy::let-unit-value`

warning: `clpy` (bin "clpy" test) generated 1 warning (run `cargo fix --bin "clpy" --tests` to apply 1 suggestion)
warning: `clpy` (bin "clpy") generated 1 warning (1 duplicate)
    Finished dev [unoptimized + debuginfo] target(s) in 0.15s
```

I expected to see this happen:

### Version

```text
rustc 1.67.0-nightly (c5d82ed7a 2022-11-19)
binary: rustc
commit-hash: c5d82ed7a4ad94a538bb87e5016e7d5ce0bd434b
commit-date: 2022-11-19
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```

**Comment by profetia** [comment]
This no longer reproducible in the latest nightly.

**Comment by ada4a** [comment]
Yes, this was fixed by https://github.com/rust-lang/rust-clippy/pull/12603

@rustbot label S-fixed
